### PR TITLE
Fix infrared sensors vs plane collision

### LINKED ISSUE #977: Wrong Collision Detection Between infra-red DistanceSensors and Plane
**Describe the Bug**
The collision detection for infra-red distance sensors is not taking into account the size of the plane:

https://github.com/cyberbotics/webots/blob/4ca4c2b3b18db7cc3cae6ce45c3db6ec930585c4/src/webots/maths/WbRay.cpp#L23
https://github.com/cyberbotics/webots/blob/4ca4c2b3b18db7cc3cae6ce45c3db6ec930585c4/src/webots/nodes/WbPlane.cpp#L345

This is easily reproducible in the following world:
```
#VRML_SIM R2019b utf8
WorldInfo {
}
Viewpoint {
  orientation -0.3490471078278512 -0.9062264631377192 -0.23857852800684398 1.1554413060032418
  position -0.5028329522780146 0.3186946910763732 0.2988871525017357
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Robot {
  children [
    DistanceSensor {
      type "infra-red"
    }
  ]
}
Transform {
  translation 0.05 1.8366e-07 -0.53
  rotation 0 0 1 1.5708
  children [
    Shape {
      appearance PBRAppearance {
        roughness 1
        metalness 0
      }
      geometry Plane {
      }
    }
  ]
}
```

We should instead implement a collision detection that takes into account the size of the plane: https://stackoverflow.com/questions/56316509/ray-bounded-plane-intersection

### PR DESCRIPTION
**Pull request by fabienrohrer**
Fix #977

### CODE REVIEW

**Comment by fabienrohrer** [comment]
👍 ✔️

**Comment by DavidMansolino** [comment]
:eyes:

File: src/webots/nodes/WbPlane.cpp [hunk 1]
```diff
@@ -340,25 +340,32 @@ double WbPlane::computeDistance(const WbRay &ray) const {
...
-  const WbTransform *const transform = upperTransform();
-  if (transform) {
-    planeNormal = transform->matrix().sub3x3MatrixDot(WbVector3(0.0, 1.0, 0.0));
-    planeNormal.normalize();
-    translation = transform->matrix().translation();
+  // 1. Compute the 4 plane vertices in world coordinates.
+  const double planeWidth = size().x();
+  const double planeHeight = size().y();
+  const WbMatrix4 &upperMatrix = upperTransform()->matrix();
+  WbVector3 p1 = upperMatrix * WbVector3(0.5 * planeWidth, 0.0, 0.5 * planeHeight);
```
Comment (DavidMansolino) [inline src/webots/nodes/WbPlane.cpp]:
```suggestion
  const WbVector3 p1 = upperMatrix * WbVector3(0.5 * planeWidth, 0.0, 0.5 * planeHeight);
```

File: src/webots/nodes/WbPlane.cpp [hunk 2]
```diff
@@ -340,25 +340,32 @@ double WbPlane::computeDistance(const WbRay &ray) const {
...
+  WbVector3 p2 = upperMatrix * WbVector3(0.5 * planeWidth, 0.0, -0.5 * planeHeight);
```
Comment (DavidMansolino) [inline src/webots/nodes/WbPlane.cpp]:
```suggestion
  const WbVector3 p2 = upperMatrix * WbVector3(0.5 * planeWidth, 0.0, -0.5 * planeHeight);
```

File: src/webots/nodes/WbPlane.cpp [hunk 3]
```diff
@@ -340,25 +340,32 @@ double WbPlane::computeDistance(const WbRay &ray) const {
...
+  WbVector3 p3 = upperMatrix * WbVector3(-0.5 * planeWidth, 0.0, -0.5 * planeHeight);
```
Comment (DavidMansolino) [inline src/webots/nodes/WbPlane.cpp]:
```suggestion
  const WbVector3 p3 = upperMatrix * WbVector3(-0.5 * planeWidth, 0.0, -0.5 * planeHeight);
```

File: src/webots/nodes/WbPlane.cpp [hunk 4]
```diff
@@ -340,25 +340,32 @@ double WbPlane::computeDistance(const WbRay &ray) const {
...
+  WbVector3 p4 = upperMatrix * WbVector3(-0.5 * planeWidth, 0.0, 0.5 * planeHeight);
```
Comment (DavidMansolino) [inline src/webots/nodes/WbPlane.cpp]:
```suggestion
  const WbVector3 p4 = upperMatrix * WbVector3(-0.5 * planeWidth, 0.0, 0.5 * planeHeight);
```

File: tests/api/controllers/distance_sensor_infra-red_vs_transformed_planes/distance_sensor_infra-red_vs_transformed_planes.c [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  for (i = 0; i < N_DISTANCE_SENSORS; ++i) {
+    ds[i] = wb_robot_get_device(name);
+    wb_distance_sensor_enable(ds[i], TIME_STEP);
+    name[0] += 1;
+  }
+
+  wb_robot_step(TIME_STEP);
+
+  for (i = 0; i < N_DISTANCE_SENSORS; ++i) {
+    double value = wb_distance_sensor_get_value(ds[i]);
```
Comment (DavidMansolino) [inline tests/api/controllers/distance_sensor_infra-red_vs_transformed_planes/distance_sensor_infra-red_vs_transformed_planes.c]:
```suggestion
    const double value = wb_distance_sensor_get_value(ds[i]);
```

File: docs/reference/changelog-r2019.md [hunk 6]
```diff
@@ -15,6 +15,7 @@ Released on ???
     - Upgraded to Qt 5.13.1 on Windows and macOS.
   - Bug fixes
     - Fixed [Lidar](lidar.md) point cloud access in controllers (thanks to Alexander).
+    - Fixed [infra-red DistanceSensors](distancesensor.md) vs [Plane](plane.md) collision detection.
```
Comment (DavidMansolino) [inline docs/reference/changelog-r2019.md]: Note that it may influence the PEN device too.

Comment (fabienrohrer) [reply docs/reference/changelog-r2019.md]: Very true, I added it in the change log.

**Comment by fabienrohrer** [comment]
👍

**Pull request merged**